File: app/src/main/java/life/plank/juna/zone/view/activity/UserProfileActivity.java
package life.plank.juna.zone.view.activity;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.gson.Gson;
import com.squareup.picasso.Picasso;

import java.net.HttpURLConnection;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.hdodenhof.circleimageview.CircleImageView;
import life.plank.juna.zone.R;
import life.plank.juna.zone.ZoneApplication;
import life.plank.juna.zone.data.model.Board;
import life.plank.juna.zone.data.model.User;
import life.plank.juna.zone.data.network.interfaces.RestApi;
import life.plank.juna.zone.util.customview.ZoneToolBar;
import life.plank.juna.zone.util.facilis.FragmentUtilKt;
import life.plank.juna.zone.view.activity.base.BaseActivity;
import life.plank.juna.zone.view.activity.home.HomeActivity;
import life.plank.juna.zone.view.adapter.GetCoinsAdapter;
import life.plank.juna.zone.view.adapter.LastTransactionsAdapter;
import life.plank.juna.zone.view.adapter.UserBoardsAdapter;
import life.plank.juna.zone.view.fragment.profile.EditProfilePopup;
import retrofit2.Response;
import retrofit2.Retrofit;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static life.plank.juna.zone.util.DataUtil.equalsNullString;
import static life.plank.juna.zone.util.DataUtil.isNullOrEmpty;
import static life.plank.juna.zone.util.PreferenceManager.getToken;
import static life.plank.juna.zone.util.facilis.FragmentUtilKt.removeActivePopupsIfAny;

public class UserProfileActivity extends BaseActivity {

    private static final String TAG = UserProfileActivity.class.getSimpleName();

    @BindView(R.id.edit_profile_button)
    Button editProfileButton;
    @BindView(R.id.profile_picture_image_view)
    CircleImageView profilePictureImageView;
    @BindView(R.id.name_text_view)
    TextView nameTextView;
    @BindView(R.id.email_text_view)
    TextView emailTextView;
    @BindView(R.id.dob_text_view)
    TextView dobTextView;
    @BindView(R.id.location_text_view)
    TextView locationTextView;
    @BindView(R.id.my_boards_list)
    RecyclerView myBoardsRecyclerView;
    @BindView(R.id.coin_count)
    TextView coinCount;
    @BindView(R.id.last_transactions_list)
    RecyclerView lastTransactionsList;
    @BindView(R.id.get_coins_list)
    RecyclerView getCoinsList;
    @BindView(R.id.settings_toolbar)
    ZoneToolBar toolbar;
    @BindView(R.id.logout_button)
    Button logoutButton;

    @Inject
    @Named("default")
    Retrofit retrofit;
    @Inject
    Gson gson;
    @Inject
    Picasso picasso;
    @Inject
    LastTransactionsAdapter lastTransactionsAdapter;
    @Inject
    GetCoinsAdapter getCoinsAdapter;
    private RestApi restApi;
    private UserBoardsAdapter userBoardsAdapter;
    private String username;

    public static void launch(Context packageContext) {
        packageContext.startActivity(new Intent(packageContext, UserProfileActivity.class));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_profile);
        ButterKnife.bind(this);

        ((ZoneApplication) getApplicationContext()).getUiComponent().inject(this);
        restApi = retrofit.create(RestApi.class);
        toolbar.setTitle(getString(R.string.settings));
        initRecyclerView();
        getUserBoards();
    }

    @Override
    protected void onResume() {
        super.onResume();
        getUserDetails();
    }

    @OnClick(R.id.edit_profile_button)
    public void editUserProfile() {
        FragmentUtilKt.pushPopup(
                getSupportFragmentManager(),
                R.id.popup_container,
                EditProfilePopup.Companion.newInstance(),
                EditProfilePopup.Companion.getTAG()
        );
    }

    private void initRecyclerView() {
        myBoardsRecyclerView.setLayoutManager(new GridLayoutManager(getApplicationContext(), 5));
        userBoardsAdapter = new UserBoardsAdapter(this, restApi, Glide.with(this));
        myBoardsRecyclerView.setAdapter(userBoardsAdapter);
        getCoinsList.setAdapter(getCoinsAdapter);
        lastTransactionsList.setAdapter(lastTransactionsAdapter);
    }

    public void getUserBoards() {
        restApi.getUserBoards(getToken())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<Response<List<Board>>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, e.getMessage());
                    }

                    @Override
                    public void onNext(Response<List<Board>> response) {
                        switch (response.code()) {
                            case HttpURLConnection.HTTP_OK:

                                List<Board> boards = response.body();
                                Board board = new Board(getString(R.string.new_));
                                if (boards != null) {
                                    boards.add(board);
                                }
                                userBoardsAdapter.setUserBoards(boards);
                                userBoardsAdapter.notifyDataSetChanged();
                                break;
                            case HttpURLConnection.HTTP_NOT_FOUND:
                                Toast.makeText(UserProfileActivity.this, R.string.cannot_find_user_boards, Toast.LENGTH_SHORT).show();
                                break;
                            default:
                                Toast.makeText(UserProfileActivity.this, R.string.something_went_wrong, Toast.LENGTH_SHORT).show();
                                break;
                        }
                    }
                });
    }

    private void getUserDetails() {
        restApi.getUser(getToken())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<Response<User>>() {
                    @Override
                    public void onCompleted() {
                        Log.i(TAG, "onCompleted");
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, e.getMessage());
                    }

                    @Override
                    public void onNext(Response<User> response) {
                        User model = response.body();
                        if (model != null) {
                            nameTextView.setText(model.getDisplayName());
                            emailTextView.setText(model.getEmailAddress());
                            username = model.getDisplayName();
                            if (model.getProfilePictureUrl() != null) {
                                picasso.load(model.getProfilePictureUrl()).into(profilePictureImageView);
                                toolbar.setProfilePic(model.getProfilePictureUrl());
                            }
                            String location;
                            if (!isNullOrEmpty(model.getCity()) && !equalsNullString(model.getCity())) {
                                location = model.getCity() + ", " + model.getCountry();
                            } else {
                                location = model.getCountry();
                            }
                            locationTextView.setText(location);
                            SharedPreferences.Editor editor = getSharedPreferences(getString(R.string.pref_user_details), MODE_PRIVATE).edit();
                            editor.putString(getString(R.string.pref_profile_pic_url), model.getProfilePictureUrl()).apply();
                            editor.putString(getString(R.string.pref_display_name), model.getDisplayName());
                        }
                    }
                });
    }

    @OnClick(R.id.home_fab)
    public void goHome() {
        HomeActivity.Companion.launch(this, true);
    }

    @Override
    public void onBackPressed() {
        if (removeActivePopupsIfAny(getSupportFragmentManager())) super.onBackPressed();
    }
}

File: app/src/main/java/life/plank/juna/zone/view/activity/profile/UserProfileActivity.kt
package life.plank.juna.zone.view.activity.profile

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.activity_user_profile.*
import kotlinx.android.synthetic.main.zone_tool_bar.*
import life.plank.juna.zone.*
import life.plank.juna.zone.data.model.Board
import life.plank.juna.zone.data.network.interfaces.RestApi
import life.plank.juna.zone.util.common.*
import life.plank.juna.zone.util.facilis.onDebouncingClick
import life.plank.juna.zone.util.sharedpreference.PreferenceManager
import life.plank.juna.zone.util.sharedpreference.PreferenceManager.Auth.getToken
import life.plank.juna.zone.util.time.DateUtil.getIsoFormattedDate
import life.plank.juna.zone.util.view.UIDisplayUtil
import life.plank.juna.zone.view.activity.base.BaseCardActivity
import life.plank.juna.zone.view.activity.home.HomeActivity
import life.plank.juna.zone.view.adapter.user.*
import life.plank.juna.zone.view.controller.BoardController
import life.plank.juna.zone.view.fragment.profile.*
import okhttp3.*
import org.jetbrains.anko.*
import org.jetbrains.anko.sdk27.coroutines.*
import java.io.File
import java.net.HttpURLConnection
import java.text.DateFormatSymbols
import java.util.*
import java.util.Calendar.*
import javax.inject.Inject


class UserProfileActivity : BaseCardActivity() {

    @Inject
    lateinit var restApi: RestApi
    @Inject
    lateinit var lastTransactionsAdapter: LastTransactionsAdapter
    @Inject
    lateinit var getCoinsAdapter: GetCoinsAdapter

    private var boardController: BoardController? = null
    private var filePath: String? = null

    companion object {
        private val TAG = UserProfileActivity::class.java.simpleName
        fun launch(from: Activity) = from.startActivity(from.intentFor<UserProfileActivity>())
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_profile)
        ZoneApplication.getApplication().uiComponent.inject(this)

        settings_toolbar.title = getString(R.string.settings)
        initRecyclerView()
        getUserBoards()

        initViews()
        setOnClickListeners()

        handleBoardIntentIfAny()
    }

    private fun initViews() {
        comment_enter_btn_switch.run {
            isChecked = PreferenceManager.App.isEnterToSend()
            text = if (isChecked) textOn else textOff
        }
    }

    override fun onResume() {
        super.onResume()
        getUserDetails()
    }

    private fun setOnClickListeners() {
        edit_profile_button.onDebouncingClick {
            pushPopup(EditProfilePopup.newInstance())
        }
        home_fab.onDebouncingClick { HomeActivity.launch(this, true) }
        profile_picture_image_view.onClick {
            if (UIDisplayUtil.checkPermission(this@UserProfileActivity)) {
                getImageResourceFromGallery()
            } else {
                customToast(R.string.add_permission)
            }
        }
        toolbar_profile_pic.onDebouncingClick {
            pushFragment(ProfileCardFragment.newInstance(), true)
        }
        comment_enter_btn_switch.onCheckedChange { _, isChecked ->
            PreferenceManager.App.saveEnterToSend(isChecked)
            comment_enter_btn_switch.text = if (isChecked) comment_enter_btn_switch.textOn else comment_enter_btn_switch.textOff
        }
    }

    private fun getImageResourceFromGallery() {
        val galleryIntent = Intent(Intent.ACTION_PICK)
        galleryIntent.setDataAndType(android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI, getString(R.string.image_format))
        startActivityForResult(galleryIntent, AppConstants.GALLERY_IMAGE_RESULT)
    }

    override fun getFragmentContainer(): Int = R.id.popup_container

    override fun restApi(): RestApi? = restApi

    private fun initRecyclerView() {
        boardController = BoardController(this, restApi, true)
        my_boards_list.setController(boardController!!)
        get_coins_list.adapter = getCoinsAdapter
        last_transactions_list.adapter = lastTransactionsAdapter
    }

    private fun getUserBoards() {
        restApi.getUserBoards(getToken()).setObserverThreadsAndSmartSubscribe({ Log.e(TAG, "getUserBoards(): ", it) }, {
            when (it.code()) {
                HttpURLConnection.HTTP_OK -> {
                    doAsync {
                        val boards = it.body() as MutableList<Board>
                        if (!boards.any { board -> board.name == getString(R.string.new_) }) {
                            boards.add(Board(getString(R.string.new_)))
                        }
                        uiThread { boardController?.setData(boards, false, null) }
                    }
                }
                HttpURLConnection.HTTP_NOT_FOUND -> errorToast(R.string.cannot_find_user_boards, it)
                else -> errorToast(R.string.something_went_wrong, it)
            }
        })
    }

    private fun getUserDetails() {
        restApi.getUser(getToken()).setObserverThreadsAndSmartSubscribe({ Log.e(TAG, "getUserDetails(): ", it) }, {
            val user = it.body()
            user?.run {
                PreferenceManager.CurrentUser.saveUser(user)
                name_text_view.text = displayName
                email_text_view.text = emailAddress
                username_text_view.text = handle

                val date = getIsoFormattedDate(dateOfBirth)
                val calendar = Calendar.getInstance()
                calendar.time = date
                val dob = "${calendar.get(DAY_OF_MONTH)} ${DateFormatSymbols().shortMonths[calendar.get(MONTH)]}, ${calendar.get(YEAR)}"
                dob_text_view.text = dob
                if (profilePictureUrl != null) {
                    Glide.with(this@UserProfileActivity).load(profilePictureUrl).into(profile_picture_image_view)
                    settings_toolbar.setProfilePic(profilePictureUrl!!)
                }
                val location: String? = if (!DataUtil.isNullOrEmpty(city) && !DataUtil.equalsNullString(city)) {
                    "$city, $country"
                } else {
                    country
                }
                PreferenceManager.CurrentUser.saveLocation(location)
                location_text_view.text = location
            }
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == AppConstants.GALLERY_IMAGE_RESULT) {
            if (resultCode == Activity.RESULT_OK) {
                try {
                    data?.data?.run {
                        filePath = UIDisplayUtil.getPathForGalleryImageView(this@run, ZoneApplication.getContext())
                        uploadProfilePicture(this@run)
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "onActivityResult()", e)
                }
            }
        }
    }

    private fun uploadProfilePicture(profilePicUri: Uri) {
        val fileToUpload = File(filePath)
        val requestBody = RequestBody.create(MediaType.parse(getString(R.string.media_type_image)), fileToUpload)
        val image = MultipartBody.Part.createFormData("", fileToUpload.name, requestBody)
        restApi.uploadProfilePicture(image, PreferenceManager.Auth.getToken())
                .setObserverThreadsAndSmartSubscribe({
                    Log.e(TAG, it.message)
                    errorToast(R.string.upload_failed, it)
                }, {
                    when (it.code()) {
                        HttpURLConnection.HTTP_NO_CONTENT -> {
                            Glide.with(this)
                                    .load(profilePicUri)
                                    .apply(RequestOptions.overrideOf(UIDisplayUtil.getDp(72f).toInt(), UIDisplayUtil.getDp(72f).toInt()))
                                    .into(profile_picture_image_view)
                            Glide.with(this).load(profilePicUri).into(toolbar_profile_pic)
                            PreferenceManager.CurrentUser.saveProfilePicUrl(it.body())
                        }
                        else -> errorToast(R.string.upload_failed, it)
                    }
                })
    }
}